src/Nethermind/Nethermind.Blockchain/TransactionPools/TransactionPool.cs

```diff
@@ -0,0 +1,218 @@
+            new ConcurrentDictionary<Type, ITransactionFilter>();
+
+        private readonly ITransactionStorage _transactionStorage;
+        private readonly IReceiptStorage _receiptStorage;
+
+        private readonly ConcurrentDictionary<PublicKey, ISynchronizationPeer> _peers =
+            new ConcurrentDictionary<PublicKey, ISynchronizationPeer>();
+
+        private readonly IEthereumSigner _signer;
+        private readonly ILogger _logger;
+        private readonly int _peerThreshold;
+
+        public TransactionPool(ITransactionStorage transactionStorage, IReceiptStorage receiptStorage,
+            IEthereumSigner signer, ILogManager logManager, int peerThreshold = 20)
+        {
+            _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+            _transactionStorage = transactionStorage;
+            _receiptStorage = receiptStorage;
+            _signer = signer;
+            _peerThreshold = peerThreshold;
+        }
+
+        public Transaction[] PendingTransactions => _pendingTransactions.Values.ToArray();
+        public TransactionReceipt GetReceipt(Keccak hash) => _receiptStorage.Get(hash);
+
+        public void AddFilter<T>(T filter) where T : ITransactionFilter
+            => _filters.TryAdd(filter.GetType(), filter);
+
+        public void DeleteFilter(Type type) => _filters.TryRemove(type, out _);
+
+        public void AddPeer(ISynchronizationPeer peer)
+        {
+            if (!_peers.TryAdd(peer.NodeId.PublicKey, peer))
+            {
+                return;
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Added peer: {peer.ClientId}");
+            }
+        }
+
+        public void DeletePeer(NodeId nodeId)
+        {
+            if (!_peers.TryRemove(nodeId.PublicKey, out _))
+            {
+                return;
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Removed peer: {nodeId}");
+            }
+        }
+
+        public void AddTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            if (_pendingTransactions.ContainsKey(transaction.Hash))
+            {
+                return;
+            }
+
+            var recoveredAddress = _signer.RecoverAddress(transaction, blockNumber);
+            if (recoveredAddress != transaction.SenderAddress)
+            {
+                throw new InvalidOperationException("Invalid signature");
+            }
+
+            _pendingTransactions.TryAdd(transaction.Hash, transaction);
+            NewPending?.Invoke(this, new TransactionEventArgs(transaction));
+            NotifyPeers(SelectPeers(GetAvailablePeers(transaction)), transaction);
+            FilterAndStoreTransaction(transaction, blockNumber);
+        }
+
+        private void FilterAndStoreTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            var filters = _filters.Values.ToArray();
```

Comment:
heap allocation for every transaction here because of the ToArray call

Reply:
Since it returns `ICollection` by default, should we use `foreach` loop instead?